docs/DesignDocs/TEECallsRefactoring.md

```diff
@@ -300,218 +282,397 @@ else
+same host cannot import the same EDL files (i.e., uinsg the same set of ECALLs). In addition,
+the proposed revision aims to remove the needs of `_u_wrapper.c` and `_t_wrapper.c`,
+which use uncommon practices (i.e., include the source file).
 
-  Because an enclave only interacts with a single host, the case of OCALL wrapper is straightforward:
-  caching the id within the implementation of the wrapper.
+To meet these goals, this document proposes a two-phase revision.
+Each phase requires separate reviews and implementation.
 
+## Phase 1: Revise the ECALL Calling Mechanism
+
+**Store ECALL by hash**
+
+The main drawback of current implementation is using the table defined
+in the `_u.c/_t.c` that serves as the contract between both ends.
+Given that the rule of importing `EDL` files is flexible, the resulting
+ECALL table is not determinstic (i.e., can be any order based on how
+EDL files are imported).
+
+To solve the problem, this phase proposes to using a single table
+that stores the function pointer along with its hash (e.g., the hash of the function name).
+Take the previous example with two enclaves, the ECALL tables now beome:
+
+- `foo` ECALL table
+  ```
+  oe_ecall_struct_t __oe_ecalls_table[] = {
+      { (oe_ecall_func_t) ecall_common_1_ecall, 3934942254 },
+      { (oe_ecall_func_t) ecall_common_2_ecall_1, 1220950296 },
+      { (oe_ecall_func_t) ecall_common_2_ecall_2, 3520030882 },
+      { (oe_ecall_func_t) ecall_foo_ecall, 2696310045 }
+  };
+  ```
+- `bar` ECALL table
+  ```
+  oe_ecall_struct_t __oe_ecalls_table[] = {
+      { (oe_ecall_func_t) ecall_common_2_ecall_1, 1220950296 },
+      { (oe_ecall_func_t) ecall_common_2_ecall_2, 3520030882 },
+      { (oe_ecall_func_t) ecall_common_1_ecall, 3934942254 },
+      { (oe_ecall_func_t) ecall_bar_ecall, 4007252781 }
+  };
   ```
-  oe_result_t oe_log_ocall(
-      uint32_t log_level,
-      const char* message)
-    {
-        oe_result_t _result = OE_FAILURE;
 
-        /* CRC32 of oe_log_ocall */
-        const uint32_t _function_hash = 3195194105;
+The the invocation of an ECALL can then be based on the hash instead of the hard-coded
+function id.
 
-        /* Used to cache the ocall id. */
-        static int _ocall_id = OE_OCALL_ID_NULL;
-        
-        ...
-        
-        /* Check if the ocall id is cached. */
-        if (_ocall_id == OE_OCALL_ID_NULL)
-          _ocall_id = oe_host_get_ocall_id_by_hash(_function_hash);
-
-        /* Call host function. */
-        if ((_result = oe_call_host_function(
-             (uint64_t) _ocall_id,
-             (uint64_t) _function_hash,
-             _input_buffer,
-             _input_buffer_size,
-             _output_buffer,
-             _output_buffer_size,
-             &_output_bytes_written)) != OE_OK)
-            goto done;
+**Register Functions**
 
-        ...
-     }
+In this phase, we keep the use of `_u_wrapper.c`/`_t_wrapper.c` that defines the registration
```

Comment:
Currently we don't have a way to ensure precise order of initialization. It is not possible with standard C/C++.
Hence if we can avoid relying on order of initialization, it would be better.

> these functnios will be invoked during the enclave initialization
(part of `_handle_init_enclave()`)

Are we going to define a new ELF section for these initialization functions?

Reply:
No. As what we did today for registering table for internal EDLs, these newly added registration functions of internal EDLs will be invoked by the host.

For the external EDL (will have a single table), the table will be passed alone with the `oe_create_enclave`, like what we did for the ocall table today.